File: app/src/main/java/com/victoria/bleled/data/remote/response/BaseResponse.java
package com.victoria.bleled.data.remote.response;

import com.victoria.bleled.data.remote.ApiException;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.List;

public class BaseResponse<T> implements Serializable {
    private int result = ApiException.SUCCESS;
    private String msg = "";
    private String reason = "";
    private T data = null;

    public static <T> BaseResponse<List<T>> convertFromResponseArray(BaseResponse<ResponseArray<T>> response) {
        BaseResponse<List<T>> resultResponse = new BaseResponse<List<T>>();
        if (response.getData() == null) {
            resultResponse.setData(new ArrayList<>());
        } else {
            resultResponse.setData(response.getData().contents);
        }
        resultResponse.setMsg(response.getMsg());
        resultResponse.setReason(response.getReason());
        resultResponse.setResult(response.getResult());

        return resultResponse;
    }

    public T getData() {
        return data;
    }

    public void setData(T data) {
        this.data = data;
    }

    public String getReason() {
        return reason;
    }

    public void setReason(String reason) {
        this.reason = reason;
    }

    public String getMsg() {
        return msg;
    }

    public void setMsg(String msg) {
        this.msg = msg;
    }

    public int getResult() {
        return result;
    }

    public void setResult(int result) {
        this.result = result;
    }
}

File: app/src/main/java/com/victoria/bleled/data/remote/myservice/BaseResponse.kt
package com.victoria.bleled.data.remote.myservice

import com.victoria.bleled.data.remote.ApiException
import java.io.Serializable
import java.util.*

class BaseResponse<T> : Serializable {
    var result: Int = ApiException.Companion.SUCCESS
    var msg = ""
    var reason = ""
    var data: T? = null
        private set

    fun setData(data: T) {
        this.data = data
    }

    companion object {
        fun <T> convertFromResponseArray(response: BaseResponse<ResponseArray<T>>): BaseResponse<List<T>> {
            val resultResponse = BaseResponse<List<T>>()
            if (response.data == null && response.data?.contents != null) {
                resultResponse.setData(ArrayList())
            } else {
                resultResponse.setData(response.data!!.contents!!)
            }
            resultResponse.msg = response.msg
            resultResponse.reason = response.reason
            resultResponse.result = response.result
            return resultResponse
        }
    }
}